aten/src/ATen/Dispatch.h

```diff
@@ -1,65 +1,71 @@
+      default:                                                           \
+        at::runtime_error(                                               \
+            "%s not implemented for '%s'", (NAME), the_type.toString()); \
+    }                                                                    \
+  }()
 
-#define DISPATCH_ALL_FLOATING_TYPES(TYPE, NAME, ...)                             \
-  [&]() {                                                                         \
-    const Type& the_type = TYPE;                                                  \
-    switch (the_type.scalarType()) {                                              \
-      case ScalarType::Double: {                                                  \
-        using scalar_t = double;                                                  \
-        return __VA_ARGS__();                                                               \
-      } case ScalarType::Float: {                                                 \
-        using scalar_t = float;                                                   \
-        return __VA_ARGS__();                                                               \
-      } case ScalarType::Half: {                                                  \
-        using scalar_t = half;                                                    \
-        return __VA_ARGS__();                                                               \
-      } default:                                                                  \
-        runtime_error("%s not implemented for '%s'", NAME, the_type.toString());  \
-    }                                                                             \
+#define AT_DISPATCH_ALL_TYPES(TYPE, NAME, ...)                           \
+  [&] {                                                                  \
+    const at::Type& the_type = TYPE;                                     \
+    switch (the_type.scalarType()) {                                     \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Byte, uint8_t, __VA_ARGS__)   \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Char, uint8_t, __VA_ARGS__)   \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Double, double, __VA_ARGS__)  \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Float, float, __VA_ARGS__)    \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Int, int32_t, __VA_ARGS__)    \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Long, int64_t, __VA_ARGS__)   \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Short, int16_t, __VA_ARGS__)  \
+      default:                                                           \
+        at::runtime_error(                                               \
+            "%s not implemented for '%s'", (NAME), the_type.toString()); \
+    }                                                                    \
   }()
 
-}
+#define AT_DISPATCH_ALL_TYPES_AND_HALF(TYPE, NAME, ...)                  \
```

Comment:
Just curious, what's the benefit of keeping it as a macro compared to a function of signature
```cpp
template<typename fn_type>
dispatch_all_types_and_half(at::Type& type, const char* name, fn_type fn) {
  ...
}
```
The template should allow us to pass in an arbitrary lambda that could capture the enclosing scope if needed. On the other hand, the implementation would be in the header, so the compiler is likely to inline the body.

Reply:
You need to specialize the lambda on the various scalar types, but lambdas can't be templated. Even C++14  generic lambdas wouldn't work well. You can do it with a struct (like the deleted `dispatch_all`), but it requires more boilerplate.